### DISCUSSION TITLE
leaflet - window is not defined

### DISCUSSION THREAD

**Discussion by pmarx92** [neuefische/questions#299] [Web] [answered]
Hello there,

i am using leaflet as map in my App. I want to use the information from the leaflet-map component in the _app.js and store them in localStorage. Now im trying to `import { MapContainer, TileLayer } from "react-leaflet";` to the _app.js in next.js but then i get the following error: 

![image](https://user-images.githubusercontent.com/74617403/199757807-d90496f1-417d-4d1a-9c3d-a3c7cf7310ef.png)

Actually, everything i want to import from react-leaflet gives me this error message. Is there a way to import those things to the _app.js? 

Commenting out this line of Code gives me the following error message:

![image](https://user-images.githubusercontent.com/74617403/199759364-898c6f23-6a54-4328-8d03-991cb9cd2eb0.png)

Now the render method of Map is the following:

![image](https://user-images.githubusercontent.com/74617403/199762365-d3ff7d77-ebf3-400e-9dfa-ac6304e1546f.png)

_app.js:
```js
import GlobalStyle from "../components/GlobalStyle";
import Layout from "../pages/Layout";
import "react-datepicker/dist/react-datepicker.css";
import useLocalStorage from "../components/hooks/LocalStorage";
import { useState } from "react";

/* import { MapContainer, TileLayer } from "react-leaflet"; */
/* import { useMapEvents } from "react-leaflet"; */

import { nanoid } from "nanoid";
import { useEffect } from "react";

function MyApp({ Component, pageProps }) {
  const [startDate, setStartDate] = useState(new Date());
  const [fishList, setFishList] = useLocalStorage("fishList", []);
  const [fishName, setFishName] = useState("");
  const [fishWeight, setFishWeight] = useState();
  const [fishLength, setFishLength] = useState();
  const [fishLocation, setFishLocation] = useState("");

  const [markerPosition, setMarkerPosition] = useLocalStorage("marker", []);
// gets the lat & lng from the map and stores them in localStorage and creates a circle onClick
  function ClickHandler() {
    const map = useMapEvents({});
    map.on("click", function (e) {
      const newMarker = {
        id: nanoid(),
        lat: e.latlng.lat,
        lng: e.latlng.lng,
      };

      L.circle([e.latlng.lat, e.latlng.lng], { radius: 200 })
        .addTo(map)
        .bindPopup("I'm a Popup!")
        .openPopup();
      setMarkerPosition([...markerPosition].concat(newMarker));
    });
    return null;
  }
//Creates a circle for every item in localStorage
  function CreateMarker() {
    const map = useMapEvents({});
    useEffect(() => {
      {
        markerPosition.map((marker) => {
          L.circle([marker.lat, marker.lng], { radius: 200 })
            .addTo(map)
            .bindPopup("I'm a Popup!")
            .openPopup();
        });
      }
    }, []);
  }

  return (
    <>
      <GlobalStyle />
      <Layout>
        <Component
          {...pageProps}
          fishList={fishList}
          setFishList={setFishList}
          startDate={startDate}
          setStartDate={setStartDate}
          fishName={fishName}
          setFishName={setFishName}
          fishWeight={fishWeight}
          setFishWeight={setFishWeight}
          fishLength={fishLength}
          setFishLength={setFishLength}
          fishLocation={fishLocation}
          setFishLocation={setFishLocation}
          ClickHandler={ClickHandler}
          CreateMarker={CreateMarker}
        />
      </Layout>
    </>
  );
}

export default MyApp;
```

**Comment by ThomasF85** [comment, marked as answer]
You cannot directly import and use anything from react-leaflet in your _app or in any page or component that is being used on the first render of a page. You should put everything that uses imports from react-leaflet in your Map component that you then import using next/dynamic. The functions you are trying to create (`ClickHandler` and `CreateMarker`) need imports from react-leaflet, so they should be inside your Map component. These functions only make sense when used on a Map, so there should be no reason to put these functions into your _app component and hand them over as props to every page.
These functions also only depend on [markerPosition, setMarkerPosition] which you read from localStorage. You should also put that into your Map component.

**Reply by pmarx92** [reply]
So far so good, but what if i want to share all the date i got from my Map to all the other components? Like, lets say the latitude and longitude?

**Reply by ThomasF85** [reply]
What data do you want to share? Latitude and longitude of your markers? You hold that information in a state that you can share without exposing anything from react-leaflet:
```
const [markerPosition, setMarkerPosition] = useLocalStorage("marker", []);
```
If your components are on other pages than the Map they can access this state exactly like this, or you can leave it in the _app.js and share it with all pages if it is needed in all pages.

**Reply by pmarx92** [reply]
Yes, i want to share the Latitude and longitude of my markers with the data i got from my form so they are "connected". I left the 
`const [markerPosition, setMarkerPosition] = useLocalStorage("marker", []);` in _app.js, then i am handing down both markerPosition and setMarkerPosition to the map:

![image](https://user-images.githubusercontent.com/74617403/199908188-f5ef6fd7-45d8-446f-94d1-582f89ca68b5.png)

handing down the useState:
![image](https://user-images.githubusercontent.com/74617403/199908243-848fd53e-c380-41d8-b631-d0d200e10592.png)

inside my map:
![image](https://user-images.githubusercontent.com/74617403/199908303-dca9bdfe-2376-4e0c-95ba-c1c20d794449.png)

then i get the following error:
![image](https://user-images.githubusercontent.com/74617403/199908378-4570667e-420d-4aff-b026-68eebc1cae63.png)

the function:
![image](https://user-images.githubusercontent.com/74617403/199908324-80309c43-789c-4ee8-8e47-5d368e75d45a.png)

I guess this error was the reason why i tried to put these functions inside the _app. Any idea how to workaround this error?

**Reply by Kheirah** [reply]
I think I can see why `markerPosition` is undefined. You are passing components `ClickHandler={ClickHandler}` and `CreateMarker={CreateMarker}` as props. And since these components expect access to `markerPosition` but haven't been given a prop themselves, your app is running into the runtime error.

In the _app.js pass down the two components as components with respective props, and you should be good to go.

**Reply by ThomasF85** [reply]
If you still run into errors, please send a link to a full version of the code. It is impossible to guess where missing props / other errors are without seeing the full code and where the error originates.

**Reply by pmarx92** [reply]
You're right, i guess it is smarter to just post the repo link.

@Kheirah I removed those functions and the props from the _app.js already, sorry i should have send the new code with the new error.

[The Map](https://github.com/pmarx92/Capstone-Project/blob/infoMarker/components/map/Map.js)
[_app.js](https://github.com/pmarx92/Capstone-Project/blob/infoMarker/pages/_app.js)

**Reply by ThomasF85** [reply]
The error is the following:
You are not passing the props `markerPosition` and `setMarkerPosition` to the Map component, that's why they are undefined in the Map component. See for example here in line 50: https://github.com/pmarx92/Capstone-Project/blob/infoMarker/pages/Add/index.js#L50
The props you are passing in _app.js get handed over to all the page components in the pages folder. These page components have to hand down the relevant props to the Map component.

**Reply by Kheirah** [reply]
This is not essential to solving the problem, but I noticed two redundant curly braces in your Map.js at lines 35 and 42 within the useEffect hook. You may remove them =)

**Reply by pmarx92** [reply]
Uff, you are right. My bad. Thank you for your patience!

@Kheirah True, i will remove them :)
